### ISSUE TITLE
Optimizing single/shared object across multiple scenes

### ISSUE THREAD

**Issue by danielkorth** [mitsuba-renderer/mitsuba3#1826]
# Summary
Trying to optimize a single texture / shared object across multiple scenes (ie different sensors, or emitters, or a rotating object). I am (1) not sure what the best/correct way to do so is and (2) I am hitting some unexpected behavior with my current approach that I don't think should happen.

My current approach is to generate multiple scenes with the desired configurations (rotated envlight & different sensor position), and then pass the same shared object (sphere) to all of them before calling `mi.load_dict()`. Then I only pass the shared object to the optimizer and it should aggregate the texture gradients across multiple views.

Unexpected Behaviors (best to commend/uncomment in the code below)
1. gradients are all zero for the texture but it is still updated/optimized
2. `dr.clip()` on the `opt[key]` prevents any optimization from happening

# Minimal repro script
```python
import glob
import os
import argparse
import mitsuba as mi
import drjit as dr
import imageio as imageio
import numpy as np
import matplotlib.pyplot as plt

parser = argparse.ArgumentParser()
parser.add_argument(
    "--output_dir",
    type=str,
    default="output_test",
    help="Directory to store output images and videos.",
)
parser.add_argument(
    "--envmap_path",
    type=str,
    default="/home/to/hdr.hdr",
    help="Path to the environment map .hdr file.",
)
parser.add_argument(
    "--texture_path",
    type=str,
    default="/home/to/texture.jpg",
    help="Path to the bitmap .jpg texture file.",
)
args = parser.parse_args()

OUTPUT_DIR = args.output_dir
ENV_MAP_PATH = args.envmap_path
TEXTURE_PATH = args.texture_path

os.makedirs(OUTPUT_DIR, exist_ok=True)

mi.set_variant('cuda_ad_rgb')

def mse(pred, gt):
    return dr.mean(dr.square(pred - gt))

# Render low noise reference images
scene = mi.load_file("mitsuba_ir/cbox.xml", res=128, integrator='prb')
log_dir = OUTPUT_DIR

def envmap_emitter(rotation_y: float = 0.0):
    return {
        'type': 'envmap',
        'filename': ENV_MAP_PATH,
        'to_world': mi.ScalarTransform4f.rotate(axis=[0, 1, 0], angle=rotation_y),
    }

def load_sensors(sensor_count: int = 4, radius: float = 5.0, phi: float = 10.0, theta: float = 30.0):
    def load_sensor(r, phi, theta):
        # Apply two rotations to convert from spherical coordinates to world 3D coordinates.
        origin = mi.ScalarTransform4f.rotate([0, 0, 1], phi).rotate([0, 1, 0], theta) @ mi.ScalarPoint3f([0, 0, r])
        return mi.load_dict({
            'type': 'perspective',
            'fov': 39.3077,
            'to_world': mi.ScalarTransform4f.look_at(
                origin=origin,
                target=[0, 0, 0],
                up=[0, 1, 0]
            ),
            'sampler': {
                'type': 'independent',
                'sample_count': 16
            },
            'film': {
                'type': 'hdrfilm',
                'width': 256,
                'height': 256,
                'rfilter': {
                    'type': 'tent',
                },
                'pixel_format': 'rgb',
            },
        })

    phis = [phi * i for i in range(sensor_count)]
    sensors = [load_sensor(radius, theta, phi) for phi in phis]
    return sensors

def load_scenes(sensor_count: int = 4):
    emitters = [envmap_emitter(rotation_y=10.0 * i) for i in range(sensor_count)]
    sensors = load_sensors(sensor_count=sensor_count)
    scenes = [
        {
            'type': 'scene',
            'integrator': {
                'type': 'prb',
                'hide_emitters': False,
                'max_depth': 3,
            },
            'shape': {
                'type': 'sphere',
                'radius': 1.0,
                'bsdf': {
                    'type': 'principled',
                    'base_color': {
                        'type': 'bitmap',
                        'filename': TEXTURE_PATH,
                    },
                    'roughness': 0.5,
                    'metallic': 0.5,
                },
            },
            'emitter': emitter,
            'sensors': sensor,
        } for emitter, sensor in zip(emitters, sensors)
    ]
    return scenes

scenes = load_scenes()

# load sphere once and pass it to all scenes
sphere = mi.load_dict(scenes[0]['shape'])
sphere_params = mi.traverse(sphere)
sphere_params['bsdf.base_color.data'] = dr.resample(sphere_params['bsdf.base_color.data'], (256, 256, 3))
for scene in scenes:
    # make sure all scenes share the same sphere instance
    scene['shape'] = sphere
    scene.update()

scenes = [mi.load_dict(scene) for scene in scenes]
image_refs = [mi.render(scene, spp=4096) for scene in scenes]
for i, image_ref in enumerate(image_refs):
    mi.util.convert_to_bitmap(image_ref)
    mi.util.write_bitmap(os.path.join(log_dir, f'image_ref_{i:02d}.png'), image_ref)

params = mi.traverse(scenes[0])
key = 'shape.bsdf.base_color.data'

# capture and save initial texture
initial_texture = dr.copy(params[key])

# generate random texture
rng = dr.rng(seed=0)
texture_opt = dr.clip(rng.random(dr.auto.TensorXf, shape=(32, 32, 3)), 0.0, 1.0)
dr.enable_grad(texture_opt)
params[key] = texture_opt

# create optimizer
opt = mi.ad.Adam(lr=0.01)
opt[key] = params[key]
params.update(opt)

losses = []
texture_errors = []
images_opt = []
for it in range(300):
    imgs = [mi.render(scene, params, spp=4) for scene in scenes]
    loss = sum([mse(img, image_ref) for img, image_ref in zip(imgs, image_refs)])

    dr.backward(loss)
    # UNEXPECTED BEHAVIOR #1: gradient is always zero despite optimization happening
    print(f"Iteration {it:02d}: non-zero gradient? {dr.any(dr.grad(params[key]) != 0)}")

    opt.step()
    # UNEXPECTED BEHAVIOR #2: no optimization happens when using clip on the opt[key] - the one below with params[key] works just fine
    # opt[key] = dr.clip(opt[key], 0.0, 1.0)
    params.update(opt)
    params[key] = dr.clip(params[key], 0.0, 1.0)

    # save images and texture for debugging
    if it % 20 == 0:
        for i, image in enumerate(imgs):
            mi.util.convert_to_bitmap(image)
            image_path = os.path.join(log_dir, f'test_{it:06d}_{i:02d}.png')
            mi.util.write_bitmap(image_path, image)
    
    if it == 100:
        opt[key] = dr.resample(opt[key], (64, 64, 3))
        params.update(opt)
    if it == 200:
        opt[key] = dr.resample(opt[key], (256, 256, 3))
        params.update(opt)
    
    # UNEXPECTED BEHAVIOR #3: texture error *sometimes* becomes 0/nan when using dr.mean/dr.square - using numpy instead works just fine
    # texture_error = dr.mean(dr.square(dr.resample(initial_texture, params[key].shape) - params[key]))
    texture_error = np.power((dr.resample(initial_texture, params[key].shape) - params[key]).numpy(), 2).mean()
    print(f'Iteration {it:02d}: loss = {loss}, texture error = {texture_error}')
    losses.append(dr.slice(loss))
    texture_errors.append(dr.slice(texture_error))

# save final renders
final_images = [mi.render(scene, params, spp=4096) for scene in scenes]
for i, image in enumerate(final_images):
    mi.util.convert_to_bitmap(image)
    mi.util.write_bitmap(os.path.join(log_dir, f'final_render_{i:02d}.png'), image)

# plot texture error curve and general loss curve
plt.plot(texture_errors)
plt.xlabel('Iteration'); plt.ylabel('Texture error'); plt.title('Texture error plot');
plt.yscale('log')
plt.savefig(os.path.join(log_dir, 'texture_error_plot.png'), dpi=150, bbox_inches='tight')
plt.close()

# save the initial and final textures
mi.util.convert_to_bitmap(params[key])
mi.util.write_bitmap(os.path.join(log_dir, 'final_texture.png'), params[key])
mi.util.convert_to_bitmap(initial_texture)
mi.util.write_bitmap(os.path.join(log_dir, 'initial_texture.png'), initial_texture)
print(f'Final loss: {loss}')

# create final texture video
final_texture_video = []
for image in sorted(glob.glob(os.path.join(log_dir, 'texture_*.png'))):
    final_texture_video.append(imageio.imread(image))
if final_texture_video:
    video_path = os.path.join(log_dir, 'final_texture_video.mp4')
    imageio.mimwrite(video_path, final_texture_video, fps=10)
    print(f'Video saved to {video_path}')

# create final video
final_video = []
for image in sorted(glob.glob(os.path.join(log_dir, 'test_*.png'))):
    final_video.append(imageio.imread(image))
if final_video:
    video_path = os.path.join(log_dir, 'final_video.mp4')
    imageio.mimwrite(video_path, final_video, fps=2)
    print(f'Video saved to {video_path}')
```

# Script Example
```
python script.py --output_dir my_outputs --envmap_path /path/to/env.hdr --texture_path /path/to/texture.jpg
```

# Environment

OS: Linux (Debian) 5.10.0-37-cloud-amd64
GPU: NVIDIA A100-SXM4-40GB
GPU Driver: 550.90.07
CUDA: 12.4
Python: 3.11
mitsuba: 3.7.1 (PyPi)
drjit: 1.2.0 (PyPi)
Variant: `cuda_ad_rgb`

**Comment by danielkorth** [comment]
Solution in issue #1493. I'm still curious to hear if my current approach for dealing with shared parameters across scenes is correct / best practice?
